File: tests/machine/x/fs/left_join_multi.fs
open System

type Anon1 = {
    id: int
    name: string
}
type Anon2 = {
    id: int
    customerId: int
}
type Anon3 = {
    orderId: int
    sku: string
}
type Anon4 = {
    orderId: obj
    name: obj
    item: obj
}
let customers: obj list = [{ id = 1; name = "Alice" }; { id = 2; name = "Bob" }]
let orders: obj list = [{ id = 100; customerId = 1 }; { id = 101; customerId = 2 }]
let items: obj list = [{ orderId = 100; sku = "a" }]
let result: obj list = [ for o in orders do 
  for c in customers do 
  let i = List.tryFind (fun i -> o.id = i.orderId) items if o.customerId = c.id then yield { orderId = o.id; name = c.name; item = i } ]
printfn "%s" "--- Left Join Multi ---"
for r in result do
    printfn "%s" (String.concat " " [string r.orderId; string r.name; string r.item])

File: tests/transpiler/x/clj/left_join_multi.clj
(ns main)

(require 'clojure.set)

(defrecord Items [orderId sku])

(defrecord Orders [id customerId])

(defrecord Customers [id name])

(def customers [{:id 1 :name "Alice"} {:id 2 :name "Bob"}])

(def orders [{:id 100 :customerId 1} {:id 101 :customerId 2}])

(def items [{:orderId 100 :sku "a"}])

(def result (for [o orders c customers i (let [i_tmp (filter (fn [i] (= (:id o) (:orderId i))) items)] (if (seq i_tmp) i_tmp [nil])) :when (= (:customerId o) (:id c))] {:orderId (:id o) :name (:name c) :item i}))

(defn -main []
  (println "--- Left Join Multi ---")
  (doseq [r result] (println (:orderId r) (:name r) (:item r))))

(-main)
